File: rust/src/solutions/s0002.rs
use proconio::input;
use std::time::Instant;

// Solution for Project Euler problem 2
// Copyright michimani All rights reserved.
//
// https://projecteuler.net/problem=2
pub fn solve() {
    let org = 4000000;
    println!("(original: {})", org);
    input! {
      limit: usize,
    }

    let start = Instant::now();

    let mut base1 = 1;
    let mut base2 = 2;
    let mut answer = 2;

    answer = loop {
        let add = base1 + base2;

        if add % 2 == 0 {
            answer += add;
        }

        base1 = base2;
        base2 = add;

        if base2 > limit {
            break answer;
        }
    };

    println!("answer is {}", answer);
    let end = start.elapsed();
    println!(
        "\nIt took {}.{:03} seconds.",
        end.as_secs(),
        end.subsec_nanos() / 1_000_000
    );
}

File: python/solution/s0002.py
#
# Solution for Project Euler problem 2
# Copyright michimani All rights reserved.
#
# https://projecteuler.net/problem=2
#


def solve():
    seq = [1, 2]
    last_num = 0
    seq_idx = 2
    ans = 2
    while last_num <= 4000000:
        pre = seq[seq_idx - 1]
        p_pre = seq[seq_idx - 2]
        last_num = pre + p_pre
        seq.append(last_num)
        if last_num % 2 == 0 and last_num <= 4000000:
            ans = ans + last_num

        seq_idx = seq_idx + 1

    return ans


if __name__ == "__main__":
    print(solve())
